docs/adr/0005-runtime-compute-slashing.md

```diff
@@ -0,0 +1,367 @@
+
+- `header.round` field is compared with the runtime's current state. If it is
+  more than `max_evidence_age` (consensus parameter) rounds behind, the evidence
+  is invalid.
+
+- Public keys of signers of both commitments are compared. If they are not the
+  same, the evidence is invalid.
+
+- Signatures of both commitments are verified. If either is invalid, the
+  evidence is invalid.
+
+- Otherwise the evidence is valid.
+
+When processing **`EquivocationBatch`** evidence, the following steps are
+performed to verify evidence validity:
+
+- The `header.round` fields of both proposed batches are compared. If they are
+  not the same, the evidence is invalid.
+
+- The `header` fields of both proposed batches are checked for basic validity.
+  If any is invalid, the evidence is invalid.
+
+- The `io_root` fields of both proposed batches are compared. If they are the
+  same, the evidence is invalid.
+
+- Public keys of signers of both commitments are compared. If they are not the
+  same, the evidence is invalid.
+
+- Signatures of both proposed batches are compared. If either is invalid, the
+  evidence is invalid.
+
+- Otherwise the evidence is valid.
+
+For all kinds of valid evidence, the following steps are performed after
+validation:
+
+- The evidence hash is derived by hashing the evidence kind and the public key
+  of the signer and the evidence is looked up in the _list of past valid
+  evidence_. If evidence already exists there, the method fails with
+  `ErrDuplicateEvidence`.
+
+- The valid evidence hash is stored in the _list of past valid evidence_.
+
+If the evidence is deemed valid by the above procedure, the misbehaving compute
+node is slashed based on the runtime slashing parameters for the
+`SlashRuntimeEquivocation` reason.
+
+Any slashing instructions related to freezing nodes are currently ignored.
+
+The node submitting the evidence may be rewarded from part of the slashed
+amount to incentivize evidence submission. The remainder of slashed funds is
+transferred to the runtime account.
+
+### Evidence Expiry
+
+On each epoch transition, for each runtime, expired evidence (as defined by the
+`max_evidence_age` and the current runtime's round) must be pruned from the
+_list of past valid evidence_.
+
+### Evidence Collection
+
+Nodes collect commitment messages distributed via the P2P gossip network and
+check for any signs of misbehavior. In case valid evidence can be constructed,
+it is submitted to the consensus layer. Any evidence parts that have expired
+should be discarded.
+
+### Consensus Parameters
+
+#### Roothash
+
+This proposal introduces the following new consensus parameters in the roothash
+module:
+
+- `max_evidence_age` (uint64) specifies the maximum age of submitted evidence in
+  the number of rounds.
+
+## Consequences
+
+### Positive
+
+- Compute nodes can be disincentivized to submit incorrect results by runtimes
+  configuring slashing parameters.
+
+### Negative
```

Comment:
Something to call out as either a negative or neutral consequence: if a runtime exhibits nondeterminism, this can result in a compute node being slashed. While we specify that runtimes should be deterministic, for non-SGX runtimes we have no way determining whether a discrepancy is due to runtime nondeterminism or a faulty compute node.

Reply:
Ok, it's probably "negative" 😄

Reply:
Added.